python/paddle/fluid/tests/unittests/op_test.py

```diff
@@ -351,13 +357,132 @@ def _calc_output(self, place, parallel=False, no_check_set=None, loss=None):
         # fetch_list = map(block.var, fetch_list)
         if not isinstance(fetch_list[0], fluid.framework.Variable):
             fetch_list = list(map(block.var, fetch_list))
+
+        if enable_inplace is not None:
+            build_strategy = fluid.BuildStrategy()
+            build_strategy.enable_inplace = enable_inplace
+            build_strategy.memory_optimize = False
+
+            compiled_prog = fluid.CompiledProgram(program).with_data_parallel(
+                build_strategy=build_strategy, places=place)
+            program = compiled_prog
+
         executor = Executor(place)
         outs = executor.run(program,
                             feed=feed_map,
                             fetch_list=fetch_list,
                             return_numpy=False)
         return outs, fetch_list
 
+    def check_inplace_output_with_place(self, place, no_check_set=None):
+        # can`t enable inplace 
+        if not fluid.core._has_infer_inplace(self.op_type):
+            return
+        expect_outs, fetch_list = self._calc_output(
+            place, no_check_set=no_check_set, enable_inplace=False)
+        actual_outs, fetch_list = self._calc_output(
+            place, no_check_set=no_check_set, enable_inplace=True)
+
+        # compare expect_outs and actual_outs
+        for i, out in enumerate(fetch_list):
+            self.assertTrue(
+                np.array_equal(
+                    np.array(expect_outs[i]), np.array(actual_outs[i])),
+                "Output (" + out.name + ") has diff at " + str(place) +
+                "when using and not using inplace")
+
+    def check_inplace_grad_output_with_place(self, place, no_check_set=None):
+        # create froward program to get forward vars
+        program = Program()
+        block = program.global_block()
+        op = self._append_ops(block)
+        inputs = self._get_inputs(block)
+        outputs = self._get_outputs(block)
+        feed_map = self.feed_var(inputs, place)
+
+        # get grad_op 
+        if not core._has_grad_op_maker(op.desc.type()):
+            return
+        grad_op_desc_list, op_grad_to_var = core.get_grad_op_desc(op.desc,
+                                                                  set(), [])
+        # has grad_op_maker but no grad_op ?
+        if not grad_op_desc_list:
+            return
+        grad_op_desc = grad_op_desc_list[0]
```

Comment:
Why only get `grad_op_desc_list[0]`? You would get all grad op descs.

Reply:
Done, thanks.